src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp

```diff
@@ -234,13 +234,13 @@ void ShenandoahHeuristics::adjust_penalty(intx step) {
 void ShenandoahHeuristics::record_success_concurrent(bool abbreviated) {
   _degenerated_cycles_in_a_row = 0;
   _successful_cycles_in_a_row++;
+  _gc_times_learned++;
 
-  if (!(abbreviated && ShenandoahAdaptiveIgnoreShortCycles)) {
+  adjust_penalty(Concurrent_Adjust);
+
+  if (_gc_times_learned <= ShenandoahLearningSteps || !(abbreviated && ShenandoahAdaptiveIgnoreShortCycles)) {
```

Comment:
Can you leave a documentation comment here recording the intent in English?

As I read it, you include the time in the recorded history if any of the following hold:
1. we haven't reached the required number of learning steps, or
2. we are not an abbreviated cycle, or
3. we haven't been asked to ignore short cycles

To the extent that this is so, I think the test would read easier if you de-Morganized the conditional into:

```
if (_gc_times_learned <= ShenandoahLearningSteps || !abbreviated || !ShenandoahAdaptiveIgnoreShortCycles)) {
```

Anyway, something for us to remember why we have this condition would be useful as well; your PR explanation is excellent. May be a condensed form of that would be appropriate as a documentation comment here?